opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -1724,6 +1723,85 @@ namespace Opm {
                 }
                 ++local_comp_index;
             }
+
+            if (handle_ms_well && !well.segments.empty()) {
+                // we need the well_ecl_ information
+                const std::string& well_name = wm.first;
+                const Well* well_ecl = getWellEcl(well_name);
+                assert(well_ecl);
+
+                const WellSegments& segment_set = well_ecl->getWellSegments(report_step);
+
+                const int top_segment_index = state.topSegmentIndex(well_index);
+                const auto& segments = well.segments;
+
+                // \Note: eventually we need to hanlde the situations that some segments are shut
+                assert(segment_set.size() == segments.size());
+
+                for (const auto& segment : segments) {
+                    const int segment_index = segment_set.segmentNumberToIndex(segment.first);
+
+                    // recovering segment rates and pressure from the restart values
+                    state.segPress()[top_segment_index + segment_index] = segment.second.pressure;
+
+                    const auto& segment_rates = segment.second.rates;
+                    for (int p = 0; p < np; ++p) {
+                        state.segRates()[(top_segment_index + segment_index) * np + p] = segment_rates.get(phs[p]);
+                    }
+                }
+            }
+        }
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    bool
+    BlackoilWellModel<TypeTag>::
+    anyMSWellOpen(const Wells* wells, const int report_step) const
+    {
+        bool any_ms_well_open = false;
+
+        const int nw = wells->number_of_wells;
+        for (int w = 0; w < nw; ++w) {
+            const std::string well_name = std::string(wells->name[w]);
+
+            const Well* well_ecl = getWellEcl(well_name);
+
+            assert(well_ecl);
+
+            if (well_ecl->isMultiSegment(report_step) ) {
+                any_ms_well_open = true;
+                break;
+            }
+        }
+        return any_ms_well_open;
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    const Well*
+    BlackoilWellModel<TypeTag>::
+    getWellEcl(const std::string& well_name) const
+    {
+        // finding the location of the well in wells_ecl
+        const int nw_wells_ecl = wells_ecl_.size();
+        int index_well = 0;
+        for (; index_well < nw_wells_ecl; ++index_well) {
```

Comment:
Looks ripe for `std::find()` and a lambda? Or is it too messy here?

Reply:
yes. I can do that. will update together with the comment above.